File: +ndi/+ontology/NDIC.m
% Location: +ndi/+ontology/NDIC.m
classdef NDIC < ndi.ontology
% NDIC - NDI Ontology object for the local NDIC controlled vocabulary file.
%   Inherits from ndi.ontology and implements lookupTermOrID for the NDIC text file.

    methods
        function obj = NDIC()
            % NDIC - Constructor for the NDIC ontology object.
            % Implicitly calls the superclass constructor ndi.ontology().
        end % constructor

        function [id, name, definition, synonyms] = lookupTermOrID(obj, term_or_id_or_name)
            % LOOKUPTERMORID - Looks up a term in the NDIC ontology file by ID or name.
            %
            %   [ID, NAME, DEFINITION, SYNONYMS] = lookupTermOrID(OBJ, TERM_OR_ID_OR_NAME)
            %
            %   Overrides the base class method to provide specific lookup
            %   functionality for the NDIC ontology using the local text file.
            %
            %   The input TERM_OR_ID_OR_NAME is the part of the original lookup string
            %   after the 'NDIC:' prefix has been removed (e.g., '8' or 'Postnatal day').
            %
            %   See also: ndi.ontology.lookup (static dispatcher)

            % Get cached data table via private static method
            ndicData = ndi.ontology.NDIC.getNDICData();
            original_input_remainder = term_or_id_or_name; % For error messages

            % Initialize Outputs
            id = ''; name = ''; definition = ''; synonyms = {};

            % Determine if input REMINDER looks like a numeric ID
            [inputNum, isNumericID] = str2num(original_input_remainder);
            % Ensure str2num succeeded and result is a scalar number
            isNumericID = isNumericID && isscalar(inputNum);

            % --- Perform Lookup ---
            rowIndex = [];
            if isNumericID
                % --- Path 1: Lookup by Numeric ID ---
                rowIndex = find(ndicData.Identifier == inputNum, 1, 'first');
                if isempty(rowIndex)
                    error('ndi:ontology:NDIC:IDNotFound', ...
                          'NDIC term with Identifier "%s" not found.', original_input_remainder);
                end
            else
                % --- Path 2: Lookup by Name (Case-Insensitive) ---
                name_to_lookup = original_input_remainder; % Remainder is the name
                match_indices = find(strcmpi(ndicData.Name, name_to_lookup));
                num_matches = numel(match_indices);

                if num_matches == 0
                    error('ndi:ontology:NDIC:NameNotFound', ...
                          'NDIC term with Name "%s" not found (case-insensitive).', name_to_lookup);
                elseif num_matches > 1
                    error('ndi:ontology:NDIC:NameNotUnique', ...
                          'Name "%s" matches multiple (%d) entries in NDIC ontology. Lookup requires a unique name.', name_to_lookup, num_matches);
                else
                    % Exactly one match found
                    rowIndex = match_indices;
                end
            end

            % --- Extract Data if Found ---
            if ~isempty(rowIndex)
                try
                    id = num2str(ndicData.Identifier(rowIndex)); % Return ID as char
                    name = char(ndicData.Name(rowIndex)); % Return canonical Name from table
                    definition = char(ndicData.Description(rowIndex)); % Return Description
                    synonyms = {}; % Always empty for NDIC
                catch ME_extract
                     error('ndi:ontology:NDIC:DataExtractionError', ...
                           'Error extracting data for input "%s" at row %d. Error: %s', original_input_remainder, rowIndex, ME_extract.message);
                end
            else
                 % This part should not be reached if logic above is correct
                 error('ndi:ontology:NDIC:InternalError', ...
                       'Internal error: Could not find index for input "%s".', original_input_remainder);
            end

        end % function lookupTermOrID

    end % methods

    methods (Static, Access = private)
        % --- Helper function to load/cache NDIC data ---
        function ndicDataTable = getNDICData()
            persistent ndicDataCache; % Cache data within this static method

            if isempty(ndicDataCache)
                fprintf('Loading NDIC ontology from file...\n');
                ontologyFilePath = ''; % Initialize
                try
                    % Use constants from base class for path components
                    ontologyFilePath = fullfile(ndi.common.PathConstants.CommonFolder,...
                        ndi.ontology.ONTOLOGY_SUBFOLDER_NDIC,...
                        ndi.ontology.NDIC_FILENAME);
                catch ME_path
                    error('ndi:ontology:NDIC:PathConstantError', ...
                          'Could not access ndi.common.PathConstants.CommonFolder. Ensure NDI paths are set up correctly. Original error: %s', ME_path.message);
                end

                if ~isfile(ontologyFilePath) % Use isfile
                    error('ndi:ontology:NDIC:FileNotFound', ...
                          'NDIC ontology file not found at: %s', ontologyFilePath);
                end
                try
                    opts = delimitedTextImportOptions('Delimiter', '\t', 'VariableNamingRule', 'preserve');
                    opts.VariableNames = {'Identifier', 'Name', 'Description'};
                    opts.VariableTypes = {'double', 'string', 'string'};
                    opts = setvartype(opts, {'Name','Description'}, 'string');
                    opts.DataLines = [2, Inf];

                    loadedData = readtable(ontologyFilePath, opts);

                    if ~all(ismember({'Identifier', 'Name', 'Description'}, loadedData.Properties.VariableNames))
                         error('ndi:ontology:NDIC:FileFormatError', ...
                               'NDIC ontology file "%s" does not contain expected columns.', ontologyFilePath);
                    end
                    if height(loadedData) == 0
                         error('ndi:ontology:NDIC:FileEmpty', ...
                               'NDIC ontology file "%s" contains no data rows.', ontologyFilePath);
                    end
                    ndicDataCache = loadedData; % Store in persistent cache
                    fprintf('NDIC ontology loaded successfully.\n');
                catch ME_read
                    ndicDataCache = []; % Clear cache on error
                    error('ndi:ontology:NDIC:FileReadError', ...
                          'Failed to read or parse NDIC ontology file "%s". Error: %s', ontologyFilePath, ME_read.message);
                end
            end
            ndicDataTable = ndicDataCache; % Return cached or newly loaded data
        end

        % --- Custom Validation Function (copied from old function) ---
        % Although input validation primarily happens in ndi.ontology.lookup now,
        % keeping it here might be useful if this class is ever instantiated directly.
        % Or it could be removed if direct instantiation is disallowed/discouraged.
        function mustBeValidInputType(input)
            if ~(isnumeric(input) && isscalar(input) || ischar(input) || (isstring(input) && isscalar(input)))
                error('ndi:ontology:NDIC:InvalidInputType', ...
                      'Input must be a scalar number, character vector, or string scalar.');
            end
        end

    end % methods (Static, Access = private)

end % classdef NDIC

File: ndi-python/ndi/ontology/ndic.py
"""
NDI Ontology NDIC - NDI Controlled Vocabulary from local file.

Inherits from Ontology and implements lookup_term_or_id for the NDIC text file.
"""

from typing import Tuple, List, Optional
import os
import pandas as pd
from .ontology import Ontology


class NDIC(Ontology):
    """
    NDIC - NDI Ontology object for the local NDIC controlled vocabulary file.

    Reads from a tab-separated file containing NDI-specific controlled vocabulary.

    File format (TSV):
        Identifier<tab>Name<tab>Description
        8<tab>Postnatal day<tab>A postnatal day...
        ...

    Examples:
        >>> ndic = NDIC()
        >>> id, name, defn, syn = ndic.lookup_term_or_id('8')
        >>> id, name, defn, syn = ndic.lookup_term_or_id('Postnatal day')
    """

    # Class-level cache for NDIC data
    _ndic_data_cache: Optional[pd.DataFrame] = None

    def __init__(self):
        """Constructor for the NDIC ontology object."""
        super().__init__()

    def lookup_term_or_id(self, term_or_id_or_name: str) -> Tuple[str, str, str, List[str]]:
        """
        Look up a term in the NDIC ontology file by ID or name.

        Args:
            term_or_id_or_name: Part after 'NDIC:' prefix has been removed
                               (e.g., '8' or 'Postnatal day')

        Returns:
            Tuple of (id, name, definition, synonyms) where:
            - id: The identifier as string
            - name: The canonical name from the table
            - definition: The description
            - synonyms: Always empty list for NDIC

        Raises:
            ValueError: If term not found or multiple matches
        """
        # Get cached data table
        ndic_data = self._get_ndic_data()
        original_input = term_or_id_or_name

        # Initialize outputs
        id_val = ''
        name = ''
        definition = ''
        synonyms = []

        # Determine if input looks like a numeric ID
        is_numeric_id = False
        input_num = None
        try:
            input_num = float(term_or_id_or_name)
            if input_num.is_integer():
                input_num = int(input_num)
                is_numeric_id = True
        except (ValueError, TypeError):
            is_numeric_id = False

        # Perform lookup
        row_index = None

        if is_numeric_id:
            # Path 1: Lookup by numeric ID
            matches = ndic_data[ndic_data['Identifier'] == input_num]
            if len(matches) == 0:
                raise ValueError(f'NDIC term with Identifier "{original_input}" not found.')
            row_index = matches.index[0]

        else:
            # Path 2: Lookup by name (case-insensitive)
            name_to_lookup = original_input
            matches = ndic_data[ndic_data['Name'].str.lower() == name_to_lookup.lower()]

            if len(matches) == 0:
                raise ValueError(f'NDIC term with Name "{name_to_lookup}" not found (case-insensitive).')
            elif len(matches) > 1:
                raise ValueError(
                    f'Name "{name_to_lookup}" matches multiple ({len(matches)}) entries in NDIC ontology. '
                    'Lookup requires a unique name.'
                )
            else:
                row_index = matches.index[0]

        # Extract data if found
        if row_index is not None:
            try:
                id_val = str(int(ndic_data.loc[row_index, 'Identifier']))  # Return ID as string
                name = str(ndic_data.loc[row_index, 'Name'])  # Return canonical name
                definition = str(ndic_data.loc[row_index, 'Description'])  # Return description
                synonyms = []  # Always empty for NDIC
            except Exception as e:
                raise ValueError(
                    f'Error extracting data for input "{original_input}" at row {row_index}: {str(e)}'
                ) from e

        return id_val, name, definition, synonyms

    @classmethod
    def _get_ndic_data(cls) -> pd.DataFrame:
        """
        Load/cache NDIC data from file.

        Returns:
            DataFrame with columns: Identifier, Name, Description

        Raises:
            FileNotFoundError: If NDIC file not found
            ValueError: If file format is invalid
        """
        if cls._ndic_data_cache is None:
            print('Loading NDIC ontology from file...')

            # Get file path using constants
            from ..common import PathConstants
            ontology_file_path = os.path.join(
                PathConstants.common_folder(),
                Ontology.ONTOLOGY_SUBFOLDER_NDIC,
                Ontology.NDIC_FILENAME
            )

            if not os.path.isfile(ontology_file_path):
                raise FileNotFoundError(f'NDIC ontology file not found at: {ontology_file_path}')

            try:
                # Read TSV file
                loaded_data = pd.read_csv(
                    ontology_file_path,
                    sep='\t',
                    dtype={'Identifier': int, 'Name': str, 'Description': str},
                    keep_default_na=False  # Don't convert empty strings to NaN
                )

                # Verify columns
                required_columns = {'Identifier', 'Name', 'Description'}
                if not required_columns.issubset(loaded_data.columns):
                    raise ValueError(
                        f'NDIC ontology file "{ontology_file_path}" does not contain expected columns: '
                        f'{required_columns}'
                    )

                if len(loaded_data) == 0:
                    raise ValueError(f'NDIC ontology file "{ontology_file_path}" contains no data rows.')

                cls._ndic_data_cache = loaded_data
                print('NDIC ontology loaded successfully.')

            except Exception as e:
                cls._ndic_data_cache = None
                raise ValueError(f'Failed to read or parse NDIC ontology file "{ontology_file_path}": {str(e)}') from e

        return cls._ndic_data_cache

    @classmethod
    def clear_cache(cls):
        """Clear the NDIC data cache."""
        cls._ndic_data_cache = None

    def __repr__(self) -> str:
        """String representation."""
        return "NDIC()"
